File: python/problem19.py

# monday = day 0
# sunday = day 6

normal_days = {
    1: 31,
    2: 28,
    3: 31,
    4: 30,
    5: 31,
    6: 30,
    7: 31,
    8: 31,
    9: 30,
    10: 31,
    11: 30,
    12: 31,
}
leap_days = {
    1: 31,
    2: 28,
    3: 31,
    4: 30,
    5: 31,
    6: 30,
    7: 31,
    8: 31,
    9: 30,
    10: 31,
    11: 30,
    12: 31,
}

def days(year):
    if year % 4 == 0:
        if year % 100 == 0:
            if year % 400 == 0:
                return leap_days
            return normal_days
        return leap_days
    return normal_days

year = 1901
day = 1

sundays = 0

while year <= 2000:
    d = days(year)
    month = 1

    while month <= 12:
        day += d[month]
        day = day % 7

        if day == 6:
            sundays += 1

        month += 1

    year += 1

print sundays

File: cpp/problem19.cpp
#include <stdio.h>
#include <stdlib.h>

using namespace std;

#define uint unsigned int

// monday = 0
// sunday = 6

uint normal_days[12] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
uint leap_days[12] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };

uint* days(uint year)
{
	if (year % 4 == 0)
	{
		if (year % 100 == 0)
		{
			if (year % 400 == 0)
			{
				return leap_days;
			}

			return normal_days;
		}

		return leap_days;
	}

	return normal_days;
}

int main(int argc, char* argv[])
{
	uint year = 1901;
	uint month;
	uint day = 1;
	uint sundays = 0;
	uint *d;

	while (year <= 2000)
	{
		d = days(year);
		month = 0;

		for (month = 0; month < 12; month++)
		{
			day += d[month];
			day %= 7;

			if (day == 6)
			{
				sundays++;
			}
		}

		year++;
	}

	printf("%u\n", sundays);
	
	return 0;
}

